### PR TITLE
Info dump 2-5/62: Moves, reviving, messages and fainting

### PR DESCRIPTION
**Pull request by End45**
It begins!

Added functions mostly related to:
- Moves and damage
- Reviving
- Message printing
- Fainting

### CODE REVIEW

**Comment by End45** [comment]
About the remaining failing checks:

- Putting EU exclusive functions at the end of the file doesn't seem right, since then they won't be in the same order as they are in the binary.
- Having to rename acronyms (specifically, "EU", "PP" and "AI") so only the first letter is uppercase looks wrong and confusing (For example, "EuFaintCheck" seems be hard to understand if you don't realize "Eu" is refers to the EU ROM acronym). Could an exception be added for those terms?

**Comment by UsernameFodder** [comment]
> About the remaining failing checks:
> 
>     * Putting EU exclusive functions at the end of the file doesn't seem right, since then they won't be in the same order as they are in the binary.
> 
>     * Having to rename acronyms (specifically, "EU", "PP" and "AI") so only the first letter is uppercase looks wrong and confusing (For example, "EuFaintCheck" seems be hard to understand if you don't realize "Eu" is refers to the EU ROM acronym). Could an exception be added for those terms?

So, these are actually both things I specifically thought about when designing the formatter 😅. I'm open to being convinced otherwise, but here were my reasons for both points:

* EU exclusive functions: It feels bad, but is there another consistent sorting scheme that could be used in general? I couldn't come up with one. Using lexicographical order (first compare NA, then EU, then JP) seemed like the most natural system to me.
  * For example, you might decide to sort based on the minimum address across all versions. But that might lead to confusing results if ordering wasn't consistent across versions.
    * E.g., you could have `symbol1 @ [NA]0x4, [EU]0x8` and `symbol2 @ [NA]0x16, [EU]0x0`. Under a "minimum" comparator, symbol2 would end up before symbol1, and scaled up to many different symbols, I don't think it would always be obvious why unless you had prior knowledge of the sorting order. This is opposed to lexicographic sorting, which I think is easy to infer by just looking at it.
    * There could also be weirdness if you have missing version information. E.g., if I add JP offsets for some symbols and not others, this might cause the symbols to end up "out of order" if just looking at, e.g., EU offsets.
  * We could try to concoct a weirder comparison scheme, but anything more complex has the risk of not being transitive, i.e. we'd have to make sure that we couldn't end up with a scenario where `A < B < C < A`. That would break sorting behavior.
* Acronym casing: I based the CamelCase convention based on [Rust's naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case), which explicitly state:
  > In UpperCamelCase, acronyms and contractions of compound words count as one word: use Uuid rather than UUID, Usize rather than USize or Stdin rather than StdIn.

  There's even a [linter check](https://rust-lang.github.io/rust-clippy/master/#upper_case_acronyms) for this for Rust source code. So it's evidently not objectively "weird" to lowercase acronyms; I think I've used things like "Pp" and "Sir" in some symbols already.
  * More generally, how many consecutive uppercase letters can you have in a name before it ceases to be CamelCase? The symbol formatter uses a limit of 2, since that's the minimum value needed to support the use of one-letter words in ACamelCase name. Growing that seems sort of arbitrary and would be less strictly catch legitimately bad names, which would make me sad; I think given the choice between two equally okay conventions, it makes more sense to choose the stricter one.

**Review by UsernameFodder — CHANGES_REQUESTED** [review]

**Comment by End45** [comment]
> > About the remaining failing checks:
> > 
> > * Putting EU exclusive functions at the end of the file doesn't seem right, since then they won't be in the same order as they are in the binary.
> > * Having to rename acronyms (specifically, "EU", "PP" and "AI") so only the first letter is uppercase looks wrong and confusing (For example, "EuFaintCheck" seems be hard to understand if you don't realize "Eu" is refers to the EU ROM acronym). Could an exception be added for those terms?
> 
> So, these are actually both things I specifically thought about when designing the formatter 😅. I'm open to being convinced otherwise, but here were my reasons for both points:
> 
>     * EU exclusive functions: It feels bad, but is there another consistent sorting scheme that could be used in general? I couldn't come up with one. Using lexicographical order (first compare NA, then EU, then JP) seemed like the most natural system to me.
>       
>       * For example, you might decide to sort based on the minimum address across all versions. But that might lead to confusing results if ordering wasn't consistent across versions. Using some other formula (e.g., the average address) would be even less obvious.
>         
>         * E.g., you could have `symbol1 @ [NA]0x4, [EU]0x4` and `symbol2 @ [EU]0x8, [JP]0x0`. Under a "minimum" comparator, symbol2 would end up before symbol1, and scaled up to many different symbols, I don't think it would always be obvious why unless you had prior knowledge of the sorting order. This is opposed to lexicographic sorting, which I think is easy to infer by just looking at it.
>         * There could also be even more weirdness if you have missing version information. E.g., if I add JP offsets for some symbols and not others, this might cause the symbols to end up "out of order" if just looking at, e.g., EU offsets.
>       * We could try to concoct a weirder comparison scheme, but anything more complex has the risk of not being transitive, i.e. we'd have to make sure that we couldn't end up with a scenario where `A < B < C < A`. That would break sorting behavior.
>         
>         * For example, an intuitive scheme like "when comparing two symbols, only compare addresses for shared versions" is nontransitive. Imagine a scenario where `symbol1 @ [NA]0x0, [EU]0x8`, `symbol2 @ [NA]0x8, [EU]0x0`, `symbol3 @ [EU]0x4`.
> 
>     * Acronym casing: I based the CamelCase convention based on [Rust's naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case), which explicitly state:
>       > In UpperCamelCase, acronyms and contractions of compound words count as one word: use Uuid rather than UUID, Usize rather than USize or Stdin rather than StdIn.
>       
>       
>       There's even a [linter check](https://rust-lang.github.io/rust-clippy/master/#upper_case_acronyms) for this for Rust source code. So it's evidently not objectively "weird" to lowercase acronyms; I think I've used things like "Pp" and "Sir" in some symbols already.
>       
>       * More generally, how many consecutive uppercase letters can you have in a name before it ceases to be CamelCase? The symbol formatter uses a limit of 2, since that's the minimum value needed to support the use of one-letter words in ACamelCase name. Growing that seems sort of arbitrary and would be less strictly catch legitimately bad names, which would make me sad; I think given the choice between two equally okay conventions, it makes more sense to choose the stricter one.

Name casing: I doubt it's possible to add exceptions for specific words so I guess I'll have to change it.

Ordering: Maybe sorting could be based on EU offsets first, US offsets next and JP offsets last? It would reduce the amount of times these ordering issues would show up since the EU ROM has the most symbols out of all versions, which means it's a better reference to know which order they should follow.
It still bugs me having to separate `SetMinimapDataE447` and `GetMinimapDataE447`, for example.

File: symbols/overlay29.yml [hunk 1]
```diff
@@ -668,6 +756,38 @@ overlay29:
...
+      address:
+        NA: 0x2303C7C
+        EU: 0x23046A8
+      description: |-
+        Makes the specified monster evolve into the specified species.
+        
+        r0: Pointer to the entity to evolve
+        r1: ?
+        r2: Species to evolve into
+    - name: ApplyDamage
```
[the remark this answers is not in the document]
Comment (End45) [reply symbols/overlay29.yml:777]: I took a look at the ASM, seems like some additional checks were introduced in EU before printing "X fainted" or "X was defeated" under some specific circumstances.
I'll add a short line noting the difference.

File: symbols/arm9.yml [hunk 2]
```diff
@@ -1801,6 +1837,90 @@ arm9:
...
+    - name: GetMonsterIdFromSpawnEntry
+      address:
+        NA: 0x2054480
+        EU: 0x20547FC
+      description: |-
+        Returns the monster ID of the specified monster spawn entry, converted to its primary gender ID (so modulo 600).
+        
+        r0: Pointer to the monster spawn entry
+        return: monster_spawn_entry::id
+    - name: GetMonsterGenderWrapper
```
[the remark this answers is not in the document]
Comment (End45) [reply symbols/arm9.yml:1915]: :+1:

File: symbols/arm9.yml [hunk 3]
```diff
@@ -1379,6 +1398,23 @@ arm9:
...
+        r0: Dungeon tip ID
+    - name: GetDungeonTipShown
+      address:
+        NA: 0x204D290
+        EU: 0x204D5C8
+      description: |-
+        Checks if a dungeon tip has already been shown before or not.
+        
+        r0: Dungeon tip ID
+        returns: True if the tip has been shown before, false otherwise.
```
[the remark this answers is not in the document]
Comment (End45) [reply symbols/arm9.yml:1417]: Fixed

File: symbols/overlay29.yml [hunk 4]
```diff
@@ -668,6 +756,38 @@ overlay29:
...
         r2: base experience gain, before boosts
+    - name: EnemyEvolution
+      address:
+        NA: 0x23026FC
+        EU: 0x2303128
+      description: |-
+        Checks if the specified enemy should evolve because it just defeated an ally, and if so, attempts to evolve it.
+        
+        r0: Pointer to the enemy to check
+    - name: Evolve
```
[the remark this answers is not in the document]
Comment (End45) [reply symbols/overlay29.yml:767]: Sure

File: headers/functions/overlay29.h [hunk 5]
```diff
@@ -42,29 +51,35 @@ bool AbilityIsActive(struct entity* entity, enum ability_id ability_id);
 bool LevitateIsActive(struct entity* entity);
 bool MonsterIsType(struct entity* entity, enum type_id type_id);
 bool IqSkillIsEnabled(struct entity* entity, enum iq_skill_id iq_id);
+enum type_id GetMoveTypeForMonster(struct entity* entity, struct move* move);
 int GetMovePower(struct entity* entity, struct move* move);
 void AddExpSpecial(struct entity* attacker, struct entity* defender, int base_exp);
+void EnemyEvolution(struct entity* enemy);
+void Evolve(struct entity* monster, undefined4* param_1, enum monster_id new_monster_id);
```
[the remark this answers is not in the document]
Comment (End45) [reply headers/functions/overlay29.h:58]: Oh, I thought they started at 0. I'll have to check all the `param_x` I've added then.

**Review by Metalcape — CHANGES_REQUESTED** [review]

File: symbols/arm9.yml [hunk 6]
```diff
@@ -1801,6 +1837,90 @@ arm9:
...
+        r1: [Output] Array that will hold the list of monster ids the specified monster can evolve into
+        r2: True to skip the check that prevents returning monsters with a different sprite size than the current one
+        r3: True to skip the check that prevents Shedinja from being counted as a potential evolution
+        return: Number of possible evolutions for the specified monster id
+    - name: GetMonsterIdFromSpawnEntry
+      address:
+        NA: 0x2054480
+        EU: 0x20547FC
+      description: |-
+        Returns the monster ID of the specified monster spawn entry, converted to its primary gender ID (so modulo 600).
```
[the remark this answers is not in the document]
Comment (End45) [reply symbols/arm9.yml:1911]: Oh, you're right, it's not 600. I'm so used to seeing that conversion that I probably didn't even check if the number was 600 or something else.
I guess I'll just remove the second part of the sentence then.

**Comment by UsernameFodder** [comment]
> Ordering: Maybe sorting could be based on EU offsets first, US offsets next and JP offsets last? It would reduce the amount of times these ordering issues would show up since the EU ROM has the most symbols out of all versions, which means it's a better reference to know which order they should follow.
> It still bugs me having to separate `SetMinimapDataE447` and `GetMinimapDataE447`, for example.

Are there any known instances of symbols in the NA version that aren't in the EU version? Or is it mostly the other way around? We could definitely reorder the version list (so that EU will be first) if it makes sense.

Now that we mostly have NA-EU parity for symbol addresses (after this PR, the only ones missing EU addresses will be `C_ROUTINES` and `OBJECTS` in overlay 11), it might actually make more sense to put EU first, since that would follow reverse-chronological order of [release date](https://bulbapedia.bulbagarden.net/wiki/Pok%C3%A9mon_Mystery_Dungeon:_Explorers_of_Sky) (EU, AU, NA, JP). That's probably why EU has the most symbols.

Regardless, please just fix the formatting here for the time being. It'll be a nightmare to rebase this PR if we reorder versions first.

**Comment by End45** [comment]
> Are there any known instances of symbols in the NA version that aren't in the EU version? Or is it mostly the other way around? We could definitely reorder the version list (so that EU will be first) if it makes sense.

According to Irdkwia's docs (which seem to have been generated using some automated tool that matches offsets for EU, US and JP), EU adds ~20 functions, removes 2 and reorders 1. So I think it would make sense to change the order.

> Regardless, please just fix the formatting here for the time being. It'll be a nightmare to rebase this PR if we reorder versions first.

Alright

**Review by UsernameFodder — APPROVED** [review]
Looks good now, thanks! I've created https://github.com/UsernameFodder/pmdsky-debug/issues/49 for switching the version order, which I'll try to handle later today.

**Pull request merged**